### PR TITLE
Import wnft

### PR DESCRIPTION
**Pull request by ruibinx**

### CODE REVIEW

File: src/pages/Wallet/NFTs/Import/index.tsx [hunk 1]
```diff
@@ -34,51 +34,56 @@ const ImportNFTModal: React.FC<{
...
-      if (parseInt(tokenId) == NaN) {
-        return -1;
-      }
-      return tokenId;
-    });
-
-    const tokenIds = await Promise.all(tokenIdPromises);
+  const getNfts = async (signer: JsonRpcSigner): Promise<Erc721[]> => {
+    const registry = new ethers.Contract(registryAddresses, RegistryABI.abi, signer);
+    const wrappedContracts: string[] = await registry.getWrappedContracts();
```
Comment (coderfengyun) [inline src/pages/Wallet/NFTs/Import/index.tsx:39]: 这个接口的设计可能不太合理，理论上应该参考ERC721Enumerable，提供wContractAtIndex的接口。否则这个list存储可能会存了太多数据。

File: src/pages/Wallet/NFTs/Import/index.tsx [hunk 2]
```diff
@@ -34,51 +34,56 @@ const ImportNFTModal: React.FC<{
...
+    console.debug('wrapped contracts', wrappedContracts);
+
+    // TODO: (ruibin) if we have a lots of wrapped contracts, performance coule be degraded.
```
Comment (coderfengyun) [inline src/pages/Wallet/NFTs/Import/index.tsx:42]: 是个问题

File: src/pages/Wallet/NFTs/Import/index.tsx [hunk 3]
```diff
@@ -141,9 +149,10 @@ const ImportNFTModal: React.FC<{
             </div>
           ) : (
             <div className={style.nftsList}>
-              {tokenData.map((item: any) => {
-                const json = Buffer.from(item.tokenURI.substring(29), 'base64').toString('utf8');
-                const result = JSON.parse(json);
+              {tokenData.map((item: Erc721) => {
+                // FIXME: (ruibin) handle different tokenUri schemes here: base64, image uri, ipfs uri.
+                // const json = Buffer.from(item.tokenURI.substring(29), 'base64').toString('utf8');
+                // const result = JSON.parse(json);
```
Comment (coderfengyun) [inline src/pages/Wallet/NFTs/Import/index.tsx:155]: Where will json and result be used ?

File: src/pages/Wallet/NFTs/Import/index.tsx [hunk 4]
```diff
@@ -34,51 +34,56 @@ const ImportNFTModal: React.FC<{
...
+          tokenInfo.push(
+            (async () => {
+              const tokenId: string = await wContract.tokenByIndex(i);
+              const tokenUri: string = await wContract.tokenURI(tokenId);
+              return {
+                contract: address,
+                tokenId: tokenId,
+                tokenURI: tokenUri,
+                name: name,
+              };
```
Comment (coderfengyun) [inline src/pages/Wallet/NFTs/Import/index.tsx:63]: Why do we need to load all the wnft from all wcontracts ?

File: src/pages/Wallet/NFTs/Import/index.tsx (hunk 3 above)
Comment (ruibinx) [reply src/pages/Wallet/NFTs/Import/index.tsx:155]: I think "json" and "result" refers to parsed NFT's ERC721 tokenURI here, which is hardcoded to only parse URI of base64 type, and use the result to display NFT image.

There are at lease three types of tokenURI scheme:

- base64
- http url
- ipfs url

We should parse URI properly, but it's not directly relevant to importing NFT, so I'm leaving a comment here.

File: src/pages/Wallet/NFTs/Import/index.tsx [hunk 5]
```diff
@@ -34,51 +34,56 @@ const ImportNFTModal: React.FC<{
...
-      const tokenId = await wrapContract?.tokenOfOwnerByIndex(Account, i);
-
-      if (parseInt(tokenId) == NaN) {
-        return -1;
-      }
-      return tokenId;
-    });
-
-    const tokenIds = await Promise.all(tokenIdPromises);
+  const getNfts = async (signer: JsonRpcSigner): Promise<Erc721[]> => {
```
[the remark this answers is not in the document]
Comment (ruibinx) [reply src/pages/Wallet/NFTs/Import/index.tsx]: done

**Pull request merged**